packages/core/functions/src/services/function-invocation-service.ts

```diff
@@ -0,0 +1,67 @@
+//
+// Copyright 2025 DXOS.org
+//
+import { Context, Effect, Layer } from 'effect';
+
+import { type SpaceId } from '@dxos/keys';
+
+import { type FunctionDefinition } from '../handler';
+
+import { FunctionImplementationResolver, LocalFunctionExecutionService } from './local-function-execution';
+import { RemoteFunctionExecutionService } from './remote-function-execution-service';
+
+export class FunctionInvocationService extends Context.Tag('@dxos/functions/FunctionInvocationService')<
+  FunctionInvocationService,
+  {
+    invokeFunction<I, O>(
+      functionDef: FunctionDefinition<I, O>,
+      input: I,
+      deployedFunctionId?: string,
+    ): Effect.Effect<O>;
+  }
+>() {
+  static layer = Layer.effect(
+    FunctionInvocationService,
+    Effect.gen(function* () {
+      const localExecutioner = yield* LocalFunctionExecutionService;
+      const remoteExecutioner = yield* RemoteFunctionExecutionService;
+
+      return {
+        invokeFunction: <I, O>(
+          functionDef: FunctionDefinition<I, O>,
+          input: I,
+          deployedFunctionId?: string,
+        ): Effect.Effect<O> =>
+          Effect.gen(function* () {
+            if (deployedFunctionId) {
+              return yield* Effect.promise(() => remoteExecutioner.callFunction(deployedFunctionId, input));
+            }
+
+            return yield* localExecutioner.invokeFunction(functionDef, input);
+          }) as Effect.Effect<O>,
```

Comment:
this cast hides an underlying problem -- the `LocalFunctionExecutor.invokeFunction` returns `Effect.Effect<unknown, never, Services>;`. Those services are not satisfied here. Update the `LocalFunctionExecutor.invokeFunction` to provide the necessary services from the layer rather then request them from the caller. Note: some service, such as Tracing, are intended to be provided by the caller and we should keep them in the signature.